### PR TITLE
Implements specific SimpleIndexReader#query method

### PR DESCRIPTION
**Pull request by tinwelint**
Instead of relying on the default wrapping found in AbstractIndexReader.
This removes unnecessary indirection in SimpleIndexReader.

### CODE REVIEW

**Comment by tinwelint** [comment]
An alternative approach to #10975 that doesn't introduce an additional `IndexReader#query` method.

**Review by pontusmelke — COMMENTED** [review]
👍

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/schema/BridgingIndexProgressor.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+ * Combine multiple progressor to act like one single logical progressor seen from client's perspective.
+ */
+public class BridgingIndexProgressor implements IndexProgressor.NodeValueClient, IndexProgressor
+{
+    private final NodeValueClient client;
+    private final int[] keys;
+    // This is a thread-safe queue because it can be used in parallel scenarios.
+    // The overhead of a concurrent queue in this case is negligable since typically there will be two or a very few number
+    // of progressors and each progressor has many results each
+    private final Queue<IndexProgressor> progressors = new ConcurrentLinkedQueue<>();
```
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/schema/BridgingIndexProgressor.java:40]: What is the parallel scenario?

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/collector/DocValuesCollector.java [hunk 2]
```diff
@@ -312,6 +320,92 @@ private void replayTo( Collector collector ) throws IOException
...
+                    if ( nextDoc != DocIdSetIterator.NO_MORE_DOCS )
+                    {
+                        index++;
+                        next = currentDocValues.get( nextDoc );
+                        return true;
+                    }
+                    else
+                    {
+                        currentIdIterator = null;
+                        return fetchNextEntityId();
```
Comment (burqen) [inline community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/collector/DocValuesCollector.java:396]: Is this recursive call really necessary? If it is, that's fine, but I would prefer something that is easier to reason about.

**Review by burqen — CHANGES_REQUESTED** [review]
Nice getting rid of that indirection! Looks good overall but got some questions.

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/collector/DocValuesCollector.java (hunk 2 above)
Comment (tinwelint) [reply community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/collector/DocValuesCollector.java:396]: This isn't changed though... so something for another PR/discussion?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/schema/BridgingIndexProgressor.java (hunk 1 above)
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/schema/BridgingIndexProgressor.java:40]: Querying a lucene index that has multiple partitions is done in parallel.

**Review by burqen — APPROVED** [review]

**Pull request merged**